### PR TITLE
User view events

### PR DESCRIPTION
**Pull request by kevinxiao27**
### Changes: 
- created user events page, events fetched from getStaticProps
- User specific registrations/saved data from users fetched on client
- EventDashboard+ EventCard components under EventDashboard; dashboard simply renders events fed in, state handling exists on the page itself. (everything is responsive with tailwind breakpoints)
- useMemo fed filter function to determine what events should be fed into the dashboard rendering component, has dependencies of ui state, text in search field and saved events.
- saved/favourited events are optimistically rendered in the UI based on user inputs regardless of the success of PATCH requests being sent to the server.

Issue: #38 

### Out of Scope (or maybe in scope actually ...):
- current backend test deployment doesn't have enough data to populate the UI (all events are unpublished, need to mock out user registrations and saved events)
- backend integration testing

### Notes:
- whole card currently links to the event registration page except clickable save button (might be a ux nightmare), do you guys have any thoughts on what should be made to be clickable vs not
- known buggy behaviour within AnimatePresence where if you spam child remounts they will not be unmounted properly. I added an artificial click delay on registered and saved buttons to prevent this from happening, but can still occur if you a key/delete repeatedly in the search bar. I confirmed this behaviour didn't have to do with the filtering logic as the events array in useMemo correctly reflects the UI state, but the motion.divs don't unmount. [Similar to this](https://github.com/framer/motion/issues/2674) but I couldn't find any workarounds on stack overflow
- mocks come from DynamoDB, faved events field doesnt exist in test but does in prod

#### Mobile
https://github.com/user-attachments/assets/1b9a96dd-b00a-4c12-888a-2d5800353bb8

#### Desktop
https://github.com/user-attachments/assets/a29a036b-8331-425c-a678-eb1f4d5163c5

- (might take a while to look through backend and mock out some events + registrations but the UI functionality is working)
- will need a follow up for backend integration

### CODE REVIEW

File: next.config.mjs [hunk 1]
```diff
@@ -1,6 +1,14 @@
 /** @type {import('next').NextConfig} */
 const nextConfig = {
   reactStrictMode: true,
+  images: {
```
[the remark this answers is not in the document]
Comment (kevinxiao27) [reply next.config.mjs:4]: Using Next.js images limits domain access, images in test data came from google. If we are hosting our own event images on the same domain we can remove this (currently it wildcard matches all hosts).

**Comment by AllanT102** [comment]
Ok couple things here that weren't in the figma design. We need two views: one for users and one for guests. Users will have registered and saved, whereas guests only see a list of the events. You can can decide what that looks like but I assume that'll just be different tabs for each. The reason why is that guests can also register for events, but they won't have an associated entry in the users table.

Secondly, to make it simple, you can just change the "registered" tab to "All events" maybe? or something like that so that there's a tab to view all events. And if the person is registered for an event, it'll just show a "REGISTERED" label instead.

What I want to see:
1. connect it with backend to query actual events
2. favourite events working
3. separate views for user and guest

How?
Spin up bt-web and create data/events that will help you test ur different cases. Make sure it is connected to ur local dynamo db instance so that when u use ur bt-webv2 to test, it has that new event data u just put in. 
Login with a user to test the view and also access it as a guest. Test out favourite events. I'm not actually too sure if the backend api for fav events works, so do some testing there as well.

Nice animations, good work so far

Let me know if you have questions.

File: src/components/EventsDashboard/EventCard.tsx [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+                  </div>
+                </div>
+                <p className="text-[10px] lg:text-sm text-events-baby-blue">
+                  {`${months[dateString.getMonth()]} ${dateString.getDate()}, ${event.year}`}, {dateString.toTimeString().slice(0, 5)}
+                </p>
+                <div className="flex flex-row items-center justify-between w-full">
+                  <p className="text-[10px] lg:text-sm text-events-baby-blue">
+                    {`${event.pricing ? "$" + event.pricing.members.toFixed(2) : "Free!"}`}{" "}
+                    {event.pricing.nonMembers ? `(Non-members ${event.pricing?.nonMembers.toFixed(2)})` : "(Members only)"}
+                  </p>
```
Comment (voctory) [inline src/components/EventsDashboard/EventCard.tsx:130]: I think code readability here could be improved by Extracting Functions, similar to how you created `registeredIndicator` and `timeStateIndicator`- what do you think

File: src/components/EventsDashboard/EventCard.tsx [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+            <div className="flex flex-row space-x-5 relative">
+              <div className="relative lg:w-[200px] lg:h-[130px] w-[100px] h-[75px] overflow-hidden rounded-lg shrink-0">
+                <Image className="object-cover" fill src={event.imageUrl} alt={event.ename} />
+              </div>
+              <div className="flex flex-col space-y-1 grow">
+                <div className="font-600 text-sm lg:text-[24px] py-0.5 lg:py-2 flex flex-row space-x-3 items-center w-full">
+                  <div>{event.ename}</div>
+                  <div className="grow"></div>
+                  <div className="hidden lg:block">{registeredIndicator(event)}</div>
+                  <div className="hidden lg:block">{timeStateIndicator(event)}</div>
```
Comment (voctory) [inline src/components/EventsDashboard/EventCard.tsx:113]: Do we need these lines? Not sure why they're hidden

File: src/pages/events.tsx [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+          <div
+            className={`bg-events-card-bg p-2 h-[46px] rounded-lg flex-row justify-center items-center space-x-1 shrink flex grow lg:hidden cursor-pointer ${
+              filterState === filterStates.saved ? "!bg-events-baby-blue" : ""
+            }`}
+            onClick={() => handleUiClick(filterStates.saved)}
+          >
+            <Bookmark height={20} width={20} className={`${filterState === filterStates.saved ? "text-events-user-card-bg fill-current" : ""}`} />
+            <p className={`${filterState === filterStates.saved ? "text-events-user-card-bg" : ""}`}>Saved</p>
+          </div>
+        </div>
```
Comment (voctory) [inline src/pages/events.tsx:152]: Two comments:
1. I was having troubles reading this code - at a glance, all of these divs look about the same, with a few minor changes. Would it be possible to improve the readability by an Extract Component refactoring?
2. Is the main difference between these divs for responsive design? Could we use Tailwind Breakpoints to simplify this further

File: src/pages/events.tsx [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+export async function getStaticProps() {
+  let events = await fetchBackend({ endpoint: "/events", method: "GET", authenticatedCall: false });
+
+  events.sort((a: BiztechEvent, b: BiztechEvent) => {
+    return new Date(b.startDate).getTime() - new Date(a.startDate).getTime();
+  });
+
+  events = events.filter((e: BiztechEvent) => {
+    return e.isPublished;
+  });
```
Comment (voctory) [inline src/pages/events.tsx:169]: Not particular to this PR, but note for @AllanT102 - in the long-term, we should be change our backend API to only expose published events for non-admins.

File: src/pages/events.tsx [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+      setSaved(user["favedEventsID;year"] ? user["favedEventsID;year"] : []);
+      setRegistered(
+        registeredEvents.data.map((event: registeredEvent) => {
+          return event["eventID;year"];
+        })
+      );
+    } catch (error) {
+      console.error(error);
+    }
+  };
```
Comment (voctory) [inline src/pages/events.tsx:69]: As Allan mentioned, we have a use-case on the old `bt-web` where people who have not yet registered for an account can browse the Events page as a Guest

**Review by voctory — COMMENTED** [review]
Hey Kevin! Great work on the mock-up implementation and animations as always. Props for the quick turnaround with the backend too

I left a similar comment about the Guest view and a few refactoring suggestions for long-term readability

Regarding your notes on the PR:
- I don't see any problem with making the whole card link to the page apart from the save button, as long as the save button is easy to click on mobile. For the clickable save button, could we use shadcn's `<Button>`? I think it would allow a bit of padding to make the icon more clickable
- Is the buggy behaviour with AnimatePresence likely? Or a rare edge-case? If it's rare, we can deprioritize finding a fix.

And about the Guest flow:
- It's hard to estimate how many of our registrations go through the Guest flow without asking the Data team, though it's likely a critical path. I'm looking forward to seeing your solution, but also please don't hesitate to message your tentative solutions to the #dev channel. The Guest flow could very well be a problem that other devs should be giving input on given the lack of Figma mock-ups, as it's more of a 'design' decision rather than something that should be figured out over a PR by a single dev

File: src/pages/events.tsx (hunk 4 above)
Comment (kevinxiao27) [reply src/pages/events.tsx:152]: Good point! I used composition before as a fast way to build the UI, but extracting components makes sense. I'll also implement breakpoints with flex-basis which should remove some of the redundant code.

File: src/pages/events.tsx [hunk 7]
```diff
@@ -15,34 +15,27 @@ interface EventProps {
...
 const filterStates = {
-  registered: "registered",
+  all: "all",
   saved: "saved",
 };
 
 export default function Page({ events }: EventProps) {
   const [searchField, setSearchField] = useState("");
-  const [filterState, setFilterState] = useState("");
+  const [filterState, setFilterState] = useState("all");
```
[the remark this answers is not in the document]
Comment (kevinxiao27) [reply src/pages/events.tsx:24]: good catch

File: src/components/EventsDashboard/EventCard.tsx (hunk 3 above)
Comment (kevinxiao27) [reply src/components/EventsDashboard/EventCard.tsx:113]: It's for the mobile design where these indicators are moved to different parts of the card (get rendered based on breakpoint). There might be a workaround using grid for it to be one-lined but I think I'll keep it like this (or we can simply keep one of the mobile/desktop placements as they don't look too different)

**Comment by kevinxiao27** [comment]
Addressed refactoring comments + commented out saved events feature from rendering in the UI, adding a ticket to fix that endpoint and uncomment this code in the future.

File: src/pages/events.tsx [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+  const [searchField, setSearchField] = useState("");
+  const [filterState, setFilterState] = useState(filterStates.all);
+  const isCooldownRef = useRef(false);
+  // these useStates will be empty arrays by default, but currently have mocks before i verify backend integration works
+  const [saved, setSaved] = useState<string[]>([]);
+  const [registered, setRegistered] = useState<string[]>([]);
+  const [email, setEmail] = useState<string>("");
+
+  useEffect(() => {
+    console.log(events);
```
Comment (AllanT102) [inline src/pages/events.tsx:36]: remove log

**Review by AllanT102 — APPROVED** [review]

**Review by AllanT102 — APPROVED** [review]

**Review by Theghost256 — APPROVED** [review]

**Pull request merged**